v2/metrics/registry.go

```diff
@@ -0,0 +1,179 @@
+		messageTypeLabel: fmt.Sprintf("%s", typeName),
+	}
+}
+
+func (m *Registry) Init(reg prom.Registerer) {
+	reg.MustRegister(
+		m.MessageReceivedCount,
+		m.MessageHandledCount,
+		m.MessageLockRenewedCount,
+		m.MessageDeadlineReachedCount,
+		m.ConcurrentMessageCount)
+}
+
+type Registry struct {
+	MessageReceivedCount        *prom.CounterVec
+	MessageHandledCount         *prom.CounterVec
+	MessageLockRenewedCount     *prom.CounterVec
+	MessageDeadlineReachedCount *prom.CounterVec
+	ConcurrentMessageCount      *prom.GaugeVec
+}
+
+// Recorder allows to initialize the metric registry and increase/decrease the registered metrics at runtime.
+type Recorder interface {
+	Init(registerer prom.Registerer)
+	IncMessageDeadlineReachedCount(msg *azservicebus.ReceivedMessage)
+	IncMessageLockRenewedFailure(msg *azservicebus.ReceivedMessage)
+	IncMessageLockRenewedSuccess(msg *azservicebus.ReceivedMessage)
+	DecConcurrentMessageCount(msg *azservicebus.ReceivedMessage)
+	IncMessageHandled(msg *azservicebus.ReceivedMessage)
+	IncMessageReceived(float64)
+	IncConcurrentMessageCount(msg *azservicebus.ReceivedMessage)
+}
+
+// IncMessageLockRenewedSuccess increase the message lock renewal success counter
+func (m *Registry) IncMessageLockRenewedSuccess(msg *azservicebus.ReceivedMessage) {
+	labels := getMessageTypeLabel(msg)
+	labels[successLabel] = "true"
+	m.MessageLockRenewedCount.With(labels).Inc()
+}
+
+// IncMessageLockRenewedFailure increase the message lock renewal failure counter
+func (m *Registry) IncMessageLockRenewedFailure(msg *azservicebus.ReceivedMessage) {
+	labels := getMessageTypeLabel(msg)
+	labels[successLabel] = "false"
+	m.MessageLockRenewedCount.With(labels).Inc()
+}
+
+// IncMessageHandled increase the message Handled
+func (m *Registry) IncMessageHandled(msg *azservicebus.ReceivedMessage) {
+	labels := getMessageTypeLabel(msg)
+	labels[deliveryCountLabel] = fmt.Sprintf("%d", msg.DeliveryCount)
+	m.MessageHandledCount.With(labels).Inc()
+}
+
+// IncConcurrentMessageCount increases the concurrent message counter
+func (m *Registry) IncConcurrentMessageCount(msg *azservicebus.ReceivedMessage) {
+	m.ConcurrentMessageCount.With(getMessageTypeLabel(msg)).Inc()
+}
+
+// DecConcurrentMessageCount decreases the concurrent message counter
+func (m *Registry) DecConcurrentMessageCount(msg *azservicebus.ReceivedMessage) {
+	m.ConcurrentMessageCount.With(getMessageTypeLabel(msg)).Dec()
+}
+
+// IncMessageDeadlineReachedCount increases the message deadline reached counter
+func (m *Registry) IncMessageDeadlineReachedCount(msg *azservicebus.ReceivedMessage) {
+	labels := getMessageTypeLabel(msg)
+	m.MessageDeadlineReachedCount.With(labels).Inc()
+}
+
+// IncMessageReceived increases the message received counter
+func (m *Registry) IncMessageReceived(count float64) {
+	m.MessageReceivedCount.With(map[string]string{}).Add(count)
+}
+
+// Informer allows to inspect metrics value stored in the registry at runtime
+type Informer struct {
```

Comment:
more of a question than feedback but - we shouldn't be using the informer outside of testing purposes according to Prometheus' conventions, right? But I'm guessing that we're doing this because at some point we'll need to know if we experience failure trying to renew a lock on a message?

Reply:
we are.
for example, it allows you to use the metrics data in the Health endpoint of your service. we do that in Fleet and in AKS. why should you not do that?